src/display/api.js

```diff
@@ -923,33 +923,50 @@ var WorkerTransport = (function WorkerTransportClosure() {
         // https://bugzilla.mozilla.org/show_bug.cgi?id=683280
-        var worker = new Worker(workerSrc);
-        var messageHandler = new MessageHandler('main', worker);
-        this.messageHandler = messageHandler;
-
-        messageHandler.on('test', function transportTest(data) {
-          var supportTypedArray = data && data.supportTypedArray;
-          if (supportTypedArray) {
-            this.worker = worker;
-            if (!data.supportTransfers) {
-              PDFJS.postMessageTransfers = false;
-            }
-            this.setupMessageHandler(messageHandler);
-            workerInitializedCapability.resolve();
-          } else {
-            this.setupFakeWorker();
-          }
-        }.bind(this));
-
-        var testObj = new Uint8Array([PDFJS.postMessageTransfers ? 255 : 0]);
-        // Some versions of Opera throw a DATA_CLONE_ERR on serializing the
-        // typed array. Also, checking if we can use transfers.
-        try {
-          messageHandler.send('test', testObj, [testObj.buffer]);
-        } catch (ex) {
-          info('Cannot use postMessage transfers');
-          testObj[0] = 0;
-          messageHandler.send('test', testObj);
+        var worker,
+            workerBackend = function(worker) { // called once worker created
+              var messageHandler = new MessageHandler('main', worker);
+              this.messageHandler = messageHandler;
+
+              messageHandler.on('test', function transportTest(data) {
+                var supportTypedArray = data && data.supportTypedArray;
+                if (supportTypedArray) {
+                  this.worker = worker;
+                  if (!data.supportTransfers) {
+                    PDFJS.postMessageTransfers = false;
+                  }
+                  this.setupMessageHandler(messageHandler);
+                  workerInitializedCapability.resolve();
+                } else {
+                  this.setupFakeWorker();
+                }
+              }.bind(this));
+
+              var testObj =
+                    new Uint8Array([PDFJS.postMessageTransfers ? 255 : 0]);
+              // Some versions of Opera throw a DATA_CLONE_ERR on serializing
+              // the typed array. Also, checking if we can use transfers.
+              try {
+                messageHandler.send('test', testObj, [testObj.buffer]);
+              } catch (ex) {
+                info('Cannot use postMessage transfers');
+                testObj[0] = 0;
+                messageHandler.send('test', testObj);
+              }
+            }.bind(this);
+
+        if (typeof workerSrc === 'function') {
+            // call the supplied function and when the src is
+            // delivered, create worker from a blob:
+            workerSrc(function(src) { // callee will callback w/ JS
+              worker =
+                new Worker(PDFJS.createObjectURL(src,
+                            'application/javascript'));
```

Comment:
What is the purpose of this patch?
You could make the API even more generic by removing the URL.createObjectURL here, and requiring the input to be a URL.